Bluetooth: Bsim: Test DFU Server phase and image index recovery from persistent storage

KKopyscinski opened the pull request:
Adds tests for DFU Server recovery on reboot. Tests are run in series, which simulates devices reboots. Each test progresses from previously reached DFU phase to next one. One additional test reaches `BT_MESH_DFU_PHASE_VERIFY_FAIL` by failing verify procedure and tests if DFU server accepts new FW Update start and can reach `BT_MESH_DFU_PHASE_APPLYING`.

This PR also contains two fixes for DFU Client and Server issues discovered while testing this behavior.

KKopyscinski committed:
tests: Bluetooth: Mesh: Test DFU Server update phase and img recovery

This adds BSim tests for DFU Server, where Firmware Distribution
procedure is stopped at given point. DFU Server is expected to save
phase and image index in persistent storage. Then new test is executed
that simulates reboot of device. Settings file is not cleared and after
setup device is expected to restore its DFU Server with correct phase
and idx.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski committed:
Bluetooth: Mesh: Do not transfer FW to targets in Verify Phase in DFU

These targets already received FW so there is no point to transfering
it again. What's more, these targets will not accept transfer and will
be dropped by DFU Client. If FW is already received DFU Client should
skip transfer and proceed to refresh procedure.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski committed:
Bluetooth: Mesh: fix FW Update procedure recovery on on DFU Server side

If device rebooted while in `BT_MESH_DFU_PHASE_TRANSFER_ACTIVE` phase
we should not recover into it, but set `BT_MESH_DFU_PHASE_TRANSFER_ERR`
instead. It's because we no longer remember which block is currently
transfered so new FW Update Start must be received and transfer
restarted.

If device rebooted while in `BT_MESH_DFU_PHASE_VERIFY_OK` it means
we didn't manage to apply image before reboot. We should enter
`BT_MESH_DFU_PHASE_VERIFY_FAIL` phase, which will allow to verify it
again and possibly apply. This is with agreement to specification
regarding receiving FW Update Start when FW is already received.

This commit also alligns handle_start with specification: when
FW is already received, we should set phase to
`BT_MESH_DFU_PHASE_VERIFY`, send Status message, then proceed to Verify.
Previously, `verify` was called before sending status, so status
contained phase set to `BT_MESH_DFU_PHASE_VERIFY_OK` or
`BT_MESH_DFU_PHASE_VERIFY_FAIL`. This would prevent DFU Client
enter refresh procedure.

`blob_recover` also expects BT_MESH_DFU_PHASE_TRANSFER_ERR now.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

PavelVPV commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -49,6 +51,46 @@ static struct bt_mesh_sar_cfg_cli sar_cfg_cli;
+
+static uint8_t stop_phase2dfu_phase(enum stop_phase phase)
+{
+	switch (phase) {
+	case DFU_SRV_STOP_NONE:
+		return BT_MESH_DFU_PHASE_IDLE;
+	case DFU_SRV_STOP_START:
+		return BT_MESH_DFU_PHASE_TRANSFER_ACTIVE;
+	case DFU_SRV_STOP_VERIFY:
+		return BT_MESH_DFU_PHASE_VERIFY;
+	case DFU_SRV_STOP_VERIFY_OK:
+		return BT_MESH_DFU_PHASE_VERIFY_OK;
+	case DFU_SRV_STOP_VERIFY_FAIL:
+		return BT_MESH_DFU_PHASE_VERIFY_FAIL;
+	case DFU_SRV_STOP_APPLY:
+		return BT_MESH_DFU_PHASE_APPLYING;
+	default:
+		return BT_MESH_DFU_PHASE_IDLE;
```
`enum stop_phase` just replicates `enum bt_mesh_dfu_phase` except some states that are unused. Could we just use it instead?

PavelVPV commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -185,10 +248,23 @@ static void target_dfu_transfer_end(struct bt_mesh_dfu_srv *srv, const struct bt
+		return;
+	}
+
 	if (dfu_verify_fail) {
 		bt_mesh_dfu_srv_rejected(srv);
+		if (expected_stop_phase == DFU_SRV_STOP_VERIFY_FAIL) {
```
Can we replace `dfu_verify_fail` with `expected_stop_phase == DFU_SRV_STOP_VERIFY_FAIL`?

PavelVPV commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -185,10 +248,23 @@ static void target_dfu_transfer_end(struct bt_mesh_dfu_srv *srv, const struct bt
 	ASSERT_TRUE(success);
 
+	if (expected_stop_phase == DFU_SRV_STOP_VERIFY) {
+		k_sem_give(&dfu_ended);
```
General comment. Would it be possible to use `dfu_ended` semaphore only in `target_dfu_apply` (on the target node) and in `dfu_cli_ended` (on the client) and use separate semaphores for each dfu step? What I don't really like now is that it becomes a test inside the test and supporting this becomes a problem. Perhaps we could do the other way around, but then the other tests need to be updated.

alxelax commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -671,6 +773,21 @@ static void cli_common_fail_on_init(void)
+{
+	const struct bt_mesh_dfu_slot *slot;
+
+	tm_set_phy_max_resync_offset(100000);
```
Why do you need this? Do you use backchannel?

KKopyscinski commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -281,6 +371,8 @@ static const struct bt_mesh_comp target_comp = {
 						BT_MESH_MODEL_CFG_CLI(&cfg_cli),
 						BT_MESH_MODEL_SAR_CFG_SRV,
 						BT_MESH_MODEL_SAR_CFG_CLI(&sar_cfg_cli),
+						BT_MESH_MODEL_CB(IMPOSTER_MODEL_ID,
```
This is needed because to fail after start we need imposter model that hijacks caps get message (like one target in fail_on_persistency). But after this I want to reboot device and setup it with composition that handles all messages properly - with normal `target_comp`. This caused problem, because there was different model count: when model settings were restored, there was no model with this index, which caused error.

PavelVPV replied:
Could you please add a comment about this?

KKopyscinski replied:
sure :)

KKopyscinski commented on subsys/bluetooth/mesh/dfu_cli.c:
```diff
@@ -432,7 +432,8 @@ static void skip_cli_from_broadcast(struct bt_mesh_dfu_cli *cli, bool skip)
 	TARGETS_FOR_EACH(cli, target) {
-		if (bt_mesh_has_addr(target->blob.addr)) {
+		if (bt_mesh_has_addr(target->blob.addr) ||
+		    target->phase == BT_MESH_DFU_PHASE_VERIFY) {
```
ok, I think `skip_targets_from_broadcast` is ok, because client case is just client being a target also.

KKopyscinski commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -764,12 +881,168 @@ static void test_cli_fail_on_persistency(void)
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(blob_id, "blob_id", NULL, blob_id_set, NULL, NULL);
```
Personally, I like leaving blob id randomization in `bt_mesh_dfu_cli_send` and using undefined value to tell it that it should be random, or passed by user otherwise. I think 0xFFFFFFFF is good marker for this. Then this will be defined and passed in DFD also. IMO this has advantage that user has also way to use random one if magic value is used. Of course user could use `sys_rand_get` oneself, but it would make things simpler

PavelVPV replied:
Ok, but then I'd suggest to use `0` as uninitialized value so that user can omit this field when initializing `struct bt_mesh_dfu_cli_xfer`. It should also be fixed in `dfd_srv.c` as currently it will have a garbage: https://github.com/zephyrproject-rtos/zephyr/blob/169270e911d3d3c699453ab0fa8618363f953913/subsys/bluetooth/mesh/dfd_srv.c#L841.

KKopyscinski replied:
ah, you want this to be part of xfer, not another argument to `bt_mesh_dfu_cli_send`. I was thinking about something like:
```c
int bt_mesh_dfu_cli_send(struct bt_mesh_dfu_cli *cli,
			 const struct bt_mesh_blob_cli_inputs *inputs,
			 const struct bt_mesh_blob_io *io,
			 const struct bt_mesh_dfu_cli_xfer *xfer,
			 uint64_t id)
{
	struct bt_mesh_dfu_target *target;

	if (bt_mesh_dfu_cli_is_busy(cli)) {
		return -EBUSY;
	}

	cli->xfer.blob.mode = xfer->mode;
	cli->xfer.blob.size = xfer->slot->size;

	if (id == BT_MESH_BLOB_ID_UNDEFINED) {
		sys_rand_get(&cli->xfer.blob.id, sizeof(cli->xfer.blob.id));
	} else {
		cli->xfer.blob.id = id;
	}
```
but I can see adding this to xfer makes it more compact. Then yes, it should be 0

KKopyscinski committed:
Bluetooth: Mesh: allow user to define BLOB ID for bt_mesh_dfu_cli_send

User can now use self-defined BLOB ID for transfer instead of using
random one each time. This allows to send multiple FW Update Start
messages with same BLOB ID, for example to resume previous one.

Signed-off-by: Krzysztof Kopyściński <krzysztof.kopyscinski@codecoup.pl>

KKopyscinski commented:
After I added configurable BLOB ID I caught additional problems:
- server in test procedure must ignore arriving chunks to be stuck in Transfer Active phase (not ignore Caps Get) so blob information will be saved in memory (Caps Get is before Xfer Start)
- To properly recover blob transfer it must expect `BT_MESH_DFU_PHASE_TRANSFER_ERR` (as this is what we restore now when rebooted during transfer)
I updated PR with these fixes too

PavelVPV commented on include/zephyr/bluetooth/mesh/dfu.h:
```diff
@@ -35,6 +35,8 @@ extern "C" {
 #define CONFIG_BT_MESH_DFU_URI_MAXLEN 0
 #endif
 
+#define BT_MESH_BLOB_ID_UNDEFINED 0x00
```
I'd rather not add it. It is only used for dfu_cli in `struct bt_mesh_dfu_cli_xfer` and description for `blob_id` parameter should be enough in my opinion.

PavelVPV commented on subsys/bluetooth/mesh/dfu_cli.c:
```diff
@@ -464,16 +465,18 @@ static void transfer(struct bt_mesh_blob_cli *b)
-		return;
-	}
-
 	/* If distributor is in the targets list, disable it until Retrieve Capabilities and BLOB
```
This comment is now relevant for `bt_mesh_has_addr()` in `skip_targets_from_broadcast`, so I would suggest to move it there.

PavelVPV commented on tests/bluetooth/bsim/mesh/src/test_dfu.c:
```diff
@@ -764,12 +859,125 @@ static void test_cli_fail_on_persistency(void)
+
+		dfu_cli.xfer.state = 5;
+		dfu_cli.xfer.slot = dfu_cli_xfer.xfer.slot;
+		dfu_cli.xfer.blob.id = blob_id;
```
Is `blob_id` initialized anywhere other than at start with zero value? Or should it be `TEST_BLOB_ID`?

KKopyscinski replied:
It should be `TEST_BLOB_ID`

PavelVPV reviewed (changes requested):
Looks very good now, just few minor comments.
